libbeat/keystore/file_keystore.go

```diff
@@ -0,0 +1,380 @@
+	return nil
+}
+
+// Delete an existing key from the store and mark the store as dirty.
+func (k *FileKeystore) Delete(key string) error {
+	k.Lock()
+	defer k.Unlock()
+
+	delete(k.secrets, key)
+	k.dirty = true
+	return nil
+}
+
+// Save persists the in memory data to disk if needed.
+func (k *FileKeystore) Save() error {
+	k.Lock()
+	err := k.doSave(true)
+	k.Unlock()
+	return err
+}
+
+// List return the availables keys.
+func (k *FileKeystore) List() ([]string, error) {
+	k.RLock()
+	defer k.RUnlock()
+
+	keys := make([]string, 0, len(k.secrets))
+	for key := range k.secrets {
+		keys = append(keys, key)
+	}
+
+	return keys, nil
+}
+
+// GetConfig returns common.Config representation of the key / secret pair to be merged with other
+// loaded configuration.
+func (k *FileKeystore) GetConfig() (*common.Config, error) {
+	k.RLock()
+	defer k.RUnlock()
+
+	configHash := make(map[string]interface{})
+	for key, secret := range k.secrets {
+		configHash[key] = string(secret.Value)
+	}
+
+	return common.NewConfigFrom(configHash)
+}
+
+// Create create an empty keystore, if the store already exist we will return an error.
+func (k *FileKeystore) Create(override bool) error {
+	k.Lock()
+	k.secrets = make(map[string]serializableSecureString)
+	k.dirty = true
+	err := k.doSave(override)
+	k.Unlock()
+	return err
+}
+
+// IsPersisted return if the keystore is physically persisted on disk.
+func (k *FileKeystore) IsPersisted() bool {
+	k.Lock()
+	defer k.Unlock()
+
+	// We just check if the file is present on disk, we don't need to do any validation
+	// for a file based keystore, since all the keys will be fetched when we initialize the object
+	// if the file is invalid it will already fails. Creating a new FileKeystore will raise
+	// any errors concerning the permissions
+	f, err := os.OpenFile(k.Path, os.O_RDONLY, filePermission)
+	if err != nil {
+		return false
+	}
+	f.Close()
+	return true
+}
+
+// doSave lock/unlocking of the ressource need to be done by the caller.
+func (k *FileKeystore) doSave(override bool) error {
+	if k.dirty == false {
+		return nil
+	}
+
+	temporaryPath := fmt.Sprintf("%s.tmp", k.Path)
+
+	w := new(bytes.Buffer)
+	jsonEncoder := json.NewEncoder(w)
+	if err := jsonEncoder.Encode(k.secrets); err != nil {
+		return fmt.Errorf("cannot serialize the keystore before saving it to disk: %v", err)
+	}
+
+	encrypted, err := k.encrypt(w)
+	if err != nil {
+		return fmt.Errorf("cannot encrypt the keystore: %v", err)
+	}
+
+	flags := os.O_RDWR | os.O_CREATE
+	if override {
+		flags |= os.O_TRUNC
+	} else {
+		flags |= os.O_EXCL
+	}
+
+	f, err := os.OpenFile(temporaryPath, flags, filePermission)
+	if err != nil {
+		return fmt.Errorf("cannot open file to save the keystore to '%s', error: %s", k.Path, err)
+	}
+
+	defer f.Close()
+
+	f.Write(version)
+	base64Encoder := base64.NewEncoder(base64.StdEncoding, f)
+	io.Copy(base64Encoder, encrypted)
+	base64Encoder.Close()
+
+	err = file.SafeFileRotate(k.Path, temporaryPath)
```

Comment:
Looks like you need to `f.Sync()` and `f.Close()` before doing the `SafeFileRotate`.

Reply:
Exactly, I've pushed the changes, but I need to configure my windows VM to be independent of Jenkins.